### DISCUSSION TITLE
pyodide.runPython(code) leads to thread blocking behaviour on first call

### DISCUSSION THREAD

**Discussion by Pipe-Runner** [pyodide/pyodide#3694] [Q&A] [answered]
Hello everyone, I am using Pyodide for a project, and everything works as expected except that for that the very time I call `pyodide.runPython(code)`, I can see a noticeable stutter in the UI (running animations are stuck for almost a second), but subsequent calls to this functions with the exact same code has no stutter at all. I have been digging around but can't find a similar case. 
My code is based on this boilerplate https://github.com/xhluca/react-pyodide-template (loads Pyodide and calls it from JS code - no worker threads or anything).
My python code is an RK4 implementation that I am using for simulating a couple of physical objects. (uses Numpy as an additional lib).

Can someone tell me if I am missing something? 

Note: My code is private until I submit my assignment, but I'll be more than happy to share snippets here if needed.

**Comment by hoodmane** [comment]
Hard to say without a reproduction.

**Comment by Pipe-Runner** [comment]
@hoodmane Thanks for taking out the time.
Here is the sandbox with my current code:

[Sanbox](https://codesandbox.io/p/github/Pipe-Runner/crown-breaker/develop?file=%2FREADME.md&workspace=%257B%2522activeFileId%2522%253A%2522clfpm998j0003g0i9h0ct1btt%2522%252C%2522openFiles%2522%253A%255B%2522%252Fsrc%252Fscreens%252FHome%252FHome.js%2522%252C%2522%252Fsrc%252Fcomponents%252FEditorPanel%252FEditorPanel.js%2522%252C%2522%252Fsrc%252Fproviders%252FPyodide%252Findex.js%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clfpm9avw000x2v6jddtgbzow%2522%253A%257B%2522key%2522%253A%2522clfpm9avw000x2v6jddtgbzow%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clfpm9avw000x2v6jddtgbzow%2522%252C%2522spacesOrder%2522%253A%255B%2522clfpm9avw000x2v6jddtgbzow%2522%255D%252C%2522hideCodeEditor%2522%253Afalse%257D)

I initialize Pyodide here:
[provider](https://codesandbox.io/p/github/Pipe-Runner/crown-breaker/develop?workspaceId=1d1b0e6d-c6d6-409e-9309-8d8f2b10225a&file=%2Fsrc%2Fproviders%2FPyodide%2Findex.js&selection=%5B%7B%22endColumn%22%3A48%2C%22endLineNumber%22%3A5%2C%22startColumn%22%3A48%2C%22startLineNumber%22%3A5%7D%5D)
and use it here:
[EditorPanel](https://codesandbox.io/p/github/Pipe-Runner/crown-breaker/develop?workspaceId=1d1b0e6d-c6d6-409e-9309-8d8f2b10225a&file=%2Fsrc%2Fcomponents%2FEditorPanel%2FEditorPanel.js&selection=%5B%7B%22endColumn%22%3A48%2C%22endLineNumber%22%3A5%2C%22startColumn%22%3A48%2C%22startLineNumber%22%3A5%7D%5D)

To reproduce the issue, just click on compute button on the side nav (opens when you click the blue button).

The python code is here:
[simulator](https://codesandbox.io/p/github/Pipe-Runner/crown-breaker/develop?workspaceId=1d1b0e6d-c6d6-409e-9309-8d8f2b10225a&file=%2Fsrc%2Fpython%2Fsimulator.py&selection=%5B%7B%22endColumn%22%3A48%2C%22endLineNumber%22%3A5%2C%22startColumn%22%3A48%2C%22startLineNumber%22%3A5%7D%5D)

Notice how you can see a stutter the fist time you click compute, but no stutter the 2nd time you click it. Not sure why.

**Comment by Pipe-Runner** [comment]
@hoodmane Sorry to bug you again, but can you please point me in the right direction? I would want to use Pyodide in my school project, but this jank is bugging me out.
I also profiled it with the Chrome dev tool, and I can see a function from Pyodide taking about 1.27 seconds blocking RAF calls. Here are the logs that you can import in your chrome dev tools to replicate the profile: https://drive.google.com/file/d/18V0xortJkX7PQYU4oa8LwNdzsr2hwsrr/view?usp=share_link

**Comment by Pipe-Runner** [comment, marked as answer]
Sorry everyone, for the unnecessary commotion. I was under the impression that the WASM code works in a non-blocking fashion, but clearly, Pyodide has more to it. The non-web worker-based solutions will lead to a non-responsive UI, and I have switched my implementation to a web-worker-based approach. Now the computations don't lead to any janks. Thanks.
